<!DOCTYPE html>
<html>
  <head><meta name="generator" content="Hexo 3.8.0">
    <meta charset="utf-8">
    <meta name="baidu-site-verification" content="vvVgSQUzPU">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <meta name="keywords" content="虎虎|博客|个人|utone|Utone|">
    <meta name="description" content="不悔梦归处 只恨太匆匆">
    
    <title>
       什么是promise -  Hexo
    </title>
    <link rel="stylesheet" href="/css/style.css"> 
    
      <link rel="icon" href="/images/15226345633649869.ico">
    
  </head>
  <body>
    <div id="container">
      <div id="main-left">
        <div id="introduce">
          <div class="logo-worp">
            <a href="/"><img class="logo" src="/images/15226345633649869.jpg" alt></a>
          </div>
          <ul>
            
              <li><a href="/"> Home </a></li>
            
              <li><a href="/archives"> Archive </a></li>
            
              <li><a href="/about"> About </a></li>
            
              <li><a href="/atom.xml"> RSS </a></li>
            
            <li>
              <img id="search-img" src="/images/search.svg" alt="search">
            </li>
          </ul>
           
            <ol class="post-toc"><li class="post-toc-item post-toc-level-3"><a class="post-toc-link" href="#Promise-是异步编程的一种解决方案：从语法上讲，promise是一个对象，从它可以获取异步操作的消息；从本意上讲，它是承诺，承诺它过一段时间会给你一个结果。promise有三种状态：pending-等待态-，fulfiled-成功态-，rejected-失败态-；状态一旦改变，就不会再变。创造promise实例后，它会立即执行。"><span class="post-toc-number">1.</span> <span class="post-toc-text">Promise 是异步编程的一种解决方案：从语法上讲，promise是一个对象，从它可以获取异步操作的消息；从本意上讲，它是承诺，承诺它过一段时间会给你一个结果。promise有三种状态：pending(等待态)，fulfiled(成功态)，rejected(失败态)；状态一旦改变，就不会再变。创造promise实例后，它会立即执行。</span></a></li><li class="post-toc-item post-toc-level-3"><a class="post-toc-link" href="#三、根据promiseA-实现一个自己的promise"><span class="post-toc-number">2.</span> <span class="post-toc-text">三、根据promiseA+实现一个自己的promise</span></a></li><li class="post-toc-item post-toc-level-3"><a class="post-toc-link" href="#步骤二：then方法链式调用"><span class="post-toc-number">3.</span> <span class="post-toc-text">步骤二：then方法链式调用</span></a></li></ol> 
          
        </div>
      </div>
      <div id="main-right"><div id="post">
    
    <header class="article-header">
        <h1>
            什么是promise
        </h1>
    </header>
    
    <div class="post-meta">
        <time class="post-data" datetime="2018-08-24T06:06:25.000Z" itemprop="datePublished">
            2018-08-24
        </time>
        
        
    </div>
    
    <div class="post-entry">
        <h3 id="Promise-是异步编程的一种解决方案：从语法上讲，promise是一个对象，从它可以获取异步操作的消息；从本意上讲，它是承诺，承诺它过一段时间会给你一个结果。promise有三种状态：pending-等待态-，fulfiled-成功态-，rejected-失败态-；状态一旦改变，就不会再变。创造promise实例后，它会立即执行。"><a href="#Promise-是异步编程的一种解决方案：从语法上讲，promise是一个对象，从它可以获取异步操作的消息；从本意上讲，它是承诺，承诺它过一段时间会给你一个结果。promise有三种状态：pending-等待态-，fulfiled-成功态-，rejected-失败态-；状态一旦改变，就不会再变。创造promise实例后，它会立即执行。" class="headerlink" title="Promise 是异步编程的一种解决方案：从语法上讲，promise是一个对象，从它可以获取异步操作的消息；从本意上讲，它是承诺，承诺它过一段时间会给你一个结果。promise有三种状态：pending(等待态)，fulfiled(成功态)，rejected(失败态)；状态一旦改变，就不会再变。创造promise实例后，它会立即执行。"></a>Promise 是异步编程的一种解决方案：从语法上讲，promise是一个对象，从它可以获取异步操作的消息；从本意上讲，它是承诺，承诺它过一段时间会给你一个结果。promise有三种状态：pending(等待态)，fulfiled(成功态)，rejected(失败态)；状态一旦改变，就不会再变。创造promise实例后，它会立即执行。</h3><p>我相信大家经常写这样的代码：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">// 当参数a大于10且参数fn2是一个方法时 执行fn2</span><br><span class="line">function fn1(a, fn2) &#123;</span><br><span class="line">    if (a &gt; 10 &amp;&amp; typeof fn2 == &apos;function&apos;) &#123;</span><br><span class="line">        fn2()</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">fn1(11, function() &#123;</span><br><span class="line">    console.log(&apos;this is a callback&apos;)</span><br><span class="line">&#125;)</span><br></pre></td></tr></table></figure>

<ul>
<li>一般来说我们会碰到的回调嵌套都不会很多，一般就一到两级，但是某些情况下，回调嵌套很多时，代码就会非常繁琐，会给我们的编程带来很多的麻烦，这种情况俗称——回调地狱。<br>这时候我们的promise就应运而生、粉墨登场了</li>
</ul>
<p>promise是用来解决两个问题的：</p>
<p>回调地狱，代码难以维护， 常常第一个的函数的输出是第二个函数的输入这种现象<br>promise可以支持多个并发的请求，获取并发请求中的数据<br>这个promise可以解决异步的问题，本身不能说promise是异步的</p>
<p>二、es6 promise用法大全<br>Promise是一个构造函数，自己身上有all、reject、resolve这几个眼熟的方法，原型上有then、catch等同样很眼熟的方法。</p>
<p>那就new一个</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">let p = new Promise((resolve, reject) =&gt; &#123;</span><br><span class="line">    //做一些异步操作</span><br><span class="line">    setTimeout(() =&gt; &#123;</span><br><span class="line">        console.log(&apos;执行完成&apos;);</span><br><span class="line">        resolve(&apos;我是成功！！&apos;);</span><br><span class="line">    &#125;, 2000);</span><br><span class="line">&#125;);</span><br></pre></td></tr></table></figure>

<p>Promise的构造函数接收一个参数：函数，并且这个函数需要传入两个参数：<br>resolve ：异步操作执行成功后的回调函数<br>reject：异步操作执行失败后的回调函数<br>then 链式操作的用法<br>所以，从表面上看，Promise只是能够简化层层回调的写法，而实质上，Promise的精髓是“状态”，用维护状态、传递状态的方式来使得回调函数能够及时调用，它比传递callback函数要简单、灵活的多。所以使用Promise的正确场景是这样的：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line">p.then((data) =&gt; &#123;</span><br><span class="line">    console.log(data);</span><br><span class="line">&#125;)</span><br><span class="line">.then((data) =&gt; &#123;</span><br><span class="line">    console.log(data);</span><br><span class="line">&#125;)</span><br><span class="line">.then((data) =&gt; &#123;</span><br><span class="line">    console.log(data);</span><br><span class="line">&#125;);</span><br><span class="line">//reject的用法 :</span><br><span class="line">//把Promise的状态置为rejected，这样我们在then中就能捕捉到，然后执行“失败”情况的回调。看下面的代码。</span><br><span class="line"></span><br><span class="line">let p = new Promise((resolve, reject) =&gt; &#123;</span><br><span class="line">    //做一些异步操作</span><br><span class="line">  setTimeout(function()&#123;</span><br><span class="line">        var num = Math.ceil(Math.random()*10); //生成1-10的随机数</span><br><span class="line">        if(num&lt;=5)&#123;</span><br><span class="line">            resolve(num);</span><br><span class="line">        &#125;</span><br><span class="line">        else&#123;</span><br><span class="line">            reject(&apos;数字太大了&apos;);</span><br><span class="line">        &#125;</span><br><span class="line">  &#125;, 2000);</span><br><span class="line">&#125;);</span><br><span class="line">p.then((data) =&gt; &#123;</span><br><span class="line">        console.log(&apos;resolved&apos;,data);</span><br><span class="line">    &#125;,(err) =&gt; &#123;</span><br><span class="line">        console.log(&apos;rejected&apos;,err);</span><br><span class="line">    &#125;</span><br><span class="line">);</span><br></pre></td></tr></table></figure>

<ul>
<li><p>then中传了两个参数，then方法可以接受两个参数，第一个对应resolve的回调，第二个对应reject的回调。所以我们能够分别拿到他们传过来的数据。多次运行这段代码，你会随机得到下面两种结果：<br>或者</p>
</li>
<li><p>catch的用法<br>我们知道Promise对象除了then方法，还有一个catch方法，它是做什么用的呢？其实它和then的第二个参数一样，用来指定reject的回调。用法是这样：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">p.then((data) =&gt; &#123;</span><br><span class="line">    console.log(&apos;resolved&apos;,data);</span><br><span class="line">&#125;).catch((err) =&gt; &#123;</span><br><span class="line">    console.log(&apos;rejected&apos;,err);</span><br><span class="line">&#125;);</span><br></pre></td></tr></table></figure>
</li>
<li><p>效果和写在then的第二个参数里面一样。不过它还有另外一个作用：在执行resolve的回调（也就是上面then中的第一个参数）时，如果抛出异常了（代码出错了），那么并不会报错卡死js，而是会进到这个catch方法中。请看下面的代码：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">p.then((data) =&gt; &#123;</span><br><span class="line">    console.log(&apos;resolved&apos;,data);</span><br><span class="line">    console.log(somedata); //此处的somedata未定义</span><br><span class="line">&#125;)</span><br><span class="line">.catch((err) =&gt; &#123;</span><br><span class="line">    console.log(&apos;rejected&apos;,err);</span><br><span class="line">&#125;);</span><br></pre></td></tr></table></figure>
</li>
<li><p>在resolve的回调中，我们console.log(somedata);而somedata这个变量是没有被定义的。如果我们不用Promise，代码运行到这里就直接在控制台报错了，不往下运行了。但是在这里，会得到这样的结果：</p>
</li>
</ul>
<ul>
<li><p>也就是说进到catch方法里面去了，而且把错误原因传到了reason参数中。即便是有错误的代码也不会报错了，这与我们的try/catch语句有相同的功能<br>all的用法：谁跑的慢，以谁为准执行回调。all接收一个数组参数，里面的值最终都算返回Promise对象<br>Promise的all方法提供了并行执行异步操作的能力，并且在所有异步操作执行完后才执行回调。看下面的例子：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">let Promise1 = new Promise(function(resolve, reject)&#123;&#125;)</span><br><span class="line">let Promise2 = new Promise(function(resolve, reject)&#123;&#125;)</span><br><span class="line">let Promise3 = new Promise(function(resolve, reject)&#123;&#125;)</span><br><span class="line"></span><br><span class="line">let p = Promise.all([Promise1, Promise2, Promise3])</span><br><span class="line"></span><br><span class="line">p.then(funciton()&#123;</span><br><span class="line">// 三个都成功则成功  </span><br><span class="line">&#125;, function()&#123;</span><br><span class="line">// 只要有失败，则失败 </span><br><span class="line">&#125;)</span><br></pre></td></tr></table></figure>
</li>
<li><p>有了all，你就可以并行执行多个异步操作，并且在一个回调中处理所有的返回数据，是不是很酷？有一个场景是很适合用这个的，一些游戏类的素材比较多的应用，打开网页时，预先加载需要用到的各种资源如图片、flash以及各种静态文件。所有的都加载完后，我们再进行页面的初始化。</p>
</li>
<li><p>race的用法：谁跑的快，以谁为准执行回调<br>race的使用场景：比如我们可以用race给某个异步请求设置超时时间，并且在超时后执行相应的操作，代码如下：<br>请求某个图片资源</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">function requestImg()&#123;</span><br><span class="line">    var p = new Promise((resolve, reject) =&gt; &#123;</span><br><span class="line">        var img = new Image();</span><br><span class="line">        img.onload = function()&#123;</span><br><span class="line">            resolve(img);</span><br><span class="line">        &#125;</span><br><span class="line">        img.src = &apos;图片的路径&apos;;</span><br><span class="line">    &#125;);</span><br><span class="line">    return p;</span><br><span class="line">&#125;</span><br><span class="line">//延时函数，用于给请求计时</span><br><span class="line">function timeout()&#123;</span><br><span class="line">    var p = new Promise((resolve, reject) =&gt; &#123;</span><br><span class="line">        setTimeout(() =&gt; &#123;</span><br><span class="line">            reject(&apos;图片请求超时&apos;);</span><br><span class="line">        &#125;, 5000);</span><br><span class="line">    &#125;);</span><br><span class="line">    return p;</span><br><span class="line">&#125;</span><br><span class="line">Promise.race([requestImg(), timeout()]).then((data) =&gt;&#123;</span><br><span class="line">    console.log(data);</span><br><span class="line">&#125;).catch((err) =&gt; &#123;</span><br><span class="line">    console.log(err);</span><br><span class="line">&#125;);</span><br></pre></td></tr></table></figure>
</li>
<li><p>requestImg函数会异步请求一张图片，我把地址写为”图片的路径”，所以肯定是无法成功请求到的。timeout函数是一个延时5秒的异步操作。我们把这两个返回Promise对象的函数放进race，于是他俩就会赛跑，如果5秒之内图片请求成功了，那么遍进入then方法，执行正常的流程。如果5秒钟图片还未成功返回，那么timeout就跑赢了，则进入catch，报出“图片请求超时”的信息。运行结果如下：</p>
</li>
</ul>
<p>好了，我相信大家对用法已经懂了，那么我们来手写一款自己的promise吧</p>
<h3 id="三、根据promiseA-实现一个自己的promise"><a href="#三、根据promiseA-实现一个自己的promise" class="headerlink" title="三、根据promiseA+实现一个自己的promise"></a>三、根据promiseA+实现一个自己的promise</h3><ul>
<li>步骤一：实现成功和失败的回调方法<br>要实现上面代码中的功能，也是promise最基本的功能。首先，需要创建一个构造函数promise，创建一个promisel类，在使用的时候传入了一个执行器executor，executor会传入两个参数：成功(resolve)和失败(reject)。之前说过，只要成功，就不会失败，只要失败就不会成功。所以，默认状态下，在调用成功时，就返回成功态，调用失败时，返回失败态。代码如下：<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br></pre></td><td class="code"><pre><span class="line">class Promise &#123;</span><br><span class="line">    constructor (executor)&#123;</span><br><span class="line">        //默认状态是等待状态</span><br><span class="line">        this.status = &apos;panding&apos;;</span><br><span class="line">        this.value = undefined;</span><br><span class="line">        this.reason = undefined;</span><br><span class="line">        //存放成功的回调</span><br><span class="line">        this.onResolvedCallbacks = [];</span><br><span class="line">        //存放失败的回调</span><br><span class="line">        this.onRejectedCallbacks = [];</span><br><span class="line">        let resolve = (data) =&gt; &#123;//this指的是实例</span><br><span class="line">            if(this.status === &apos;pending&apos;)&#123;</span><br><span class="line">                this.value = data;</span><br><span class="line">                this.status = &quot;resolved&quot;;</span><br><span class="line">                this.onResolvedCallbacks.forEach(fn =&gt; fn());</span><br><span class="line">            &#125;</span><br><span class="line"> </span><br><span class="line">        &#125;</span><br><span class="line">        let reject = (reason) =&gt; &#123;</span><br><span class="line">            if(this.status === &apos;pending&apos;)&#123;</span><br><span class="line">                this.reason = reason;</span><br><span class="line">                this.status = &apos;rejected&apos;;</span><br><span class="line">                this.onRejectedCallbacks.forEach(fn =&gt; fn());</span><br><span class="line">            &#125;</span><br><span class="line">        &#125;</span><br><span class="line">        try&#123;//执行时可能会发生异常</span><br><span class="line">            executor(resolve,reject);</span><br><span class="line">        &#125;catch (e)&#123;</span><br><span class="line">            reject(e);//promise失败了</span><br><span class="line">        &#125;</span><br><span class="line">       </span><br><span class="line">    &#125;    </span><br><span class="line">//promise A+规范规定，在有异常错误时，则执行失败函数。</span><br><span class="line"></span><br><span class="line">constructor (executor)&#123;</span><br><span class="line">    ......      try&#123;</span><br><span class="line">        executor(resolve,reject);</span><br><span class="line">      &#125;catch(e)&#123;</span><br><span class="line">        reject(e);</span><br><span class="line">      &#125;</span><br><span class="line">  &#125;</span><br></pre></td></tr></table></figure>

</li>
</ul>
<h3 id="步骤二：then方法链式调用"><a href="#步骤二：then方法链式调用" class="headerlink" title="步骤二：then方法链式调用"></a>步骤二：then方法链式调用</h3><p>then方法是promise的最基本的方法，返回的是两个回调，一个成功的回调，一个失败的回调，实现过程如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">  then(onFulFilled, onRejected) &#123;</span><br><span class="line">    if (this.status === &apos;resolved&apos;) &#123; //成功状态的回调</span><br><span class="line">      onFulFilled(this.value);</span><br><span class="line">    &#125;</span><br><span class="line">    if (this.status === &apos;rejected&apos;) &#123;//失败状态的回调</span><br><span class="line">      onRejected(this.reason);</span><br><span class="line">    &#125;</span><br><span class="line">  &#125;</span><br><span class="line">let p = new Promise(function()&#123;</span><br><span class="line">    resolve(&apos;我是成功&apos;);</span><br><span class="line">&#125;)</span><br><span class="line">        p.then((data) =&gt; &#123;console.log(data);&#125;,(err) =&gt; &#123;&#125;);</span><br><span class="line">        p.then((data) =&gt; &#123;console.log(data);&#125;,(err) =&gt; &#123;&#125;);</span><br><span class="line">        p.then((data) =&gt; &#123;console.log(data);&#125;,(err) =&gt; &#123;&#125;);</span><br><span class="line">        返回的结果是：</span><br><span class="line"></span><br><span class="line">        我是成功</span><br><span class="line">        我是成功</span><br><span class="line">        我是成功</span><br></pre></td></tr></table></figure>

<ul>
<li>为了实现这样的效果，则上一次的代码将要重新写过，我们可以把每次调用resolve的结果存入一个数组中，每次调用reject的结果存入一个数组。这就是为何会在上面定义两个数组,且分别在resolve()和reject()遍历两个数组的原因。因此，在调用resolve()或者reject()之前，我们在pending状态时，会把多次then中的结果存入数组中，则上面的代码会改变为：<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line">  then(onFulFilled, onRejected) &#123;</span><br><span class="line">    if (this.status === &apos;resolved&apos;) &#123;</span><br><span class="line">      onFulFilled(this.value);</span><br><span class="line">    &#125;</span><br><span class="line">    if (this.status === &apos;rejected&apos;) &#123;</span><br><span class="line">      onRejected(this.reason);</span><br><span class="line">    &#125;</span><br><span class="line">    // 当前既没有完成 也没有失败</span><br><span class="line">    if (this.status === &apos;pending&apos;) &#123;</span><br><span class="line">      // 存放成功的回调</span><br><span class="line">      this.onResolvedCallbacks.push(() =&gt; &#123;</span><br><span class="line">        onFulFilled(this.value);</span><br><span class="line">      &#125;);</span><br><span class="line">      // 存放失败的回调</span><br><span class="line">      this.onRejectedCallbacks.push(() =&gt; &#123;</span><br><span class="line">        onRejected(this.reason);</span><br><span class="line">      &#125;);</span><br><span class="line">    &#125;</span><br><span class="line">  &#125;</span><br><span class="line"> </span><br><span class="line">Promise A+规范中规定then方法可以链式调用</span><br><span class="line">在promise中，要实现链式调用返回的结果是返回一个新的promise，第一次then中返回的结果，无论是成功或失败，都将返回到下一次then中的成功态中，但在第一次then中如果抛出异常错误，则将返回到下一次then中的失败态中</span><br><span class="line"></span><br><span class="line">链式调用成功时</span><br><span class="line"></span><br><span class="line">链式调用成功会返回值，有多种情况，根据举的例子，大致列出可能会发生的结果。因此将链式调用返回的值单独写一个方法。方法中传入四个参数，分别是p2,x,resolve,reject,p2指的是上一次返回的promise，x表示运行promise返回的结果，resolve和reject是p2的方法。则代码写为：</span><br><span class="line">function resolvePromise(p2,x,resolve,reject)&#123;</span><br><span class="line">    ....</span><br><span class="line">&#125;</span><br><span class="line">返回结果不能是自己</span><br></pre></td></tr></table></figure>

</li>
</ul>
<p>var p = new Promise((resovle,reject) =&gt; {<br>    return p;     //返回的结果不能是自己，<br>})</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">当返回结果是自己时，永远也不会成功或失败，因此当返回自己时，应抛出一个错误</span><br></pre></td></tr></table></figure>

<p>function resolvePromise(p2,x,resolve,reject){<br>    if(px===x){<br>        return reject(new TypeError(‘自己引用自己了’));<br>    }<br>    ….<br>}
返回结果可能是promise<br>function resolvePromise(promise2,x,resolve,reject){<br>    //判断x是不是promise<br>    //规范中规定：我们允许别人乱写，这个代码可以实现我们的promise和别人的promise 进行交互<br>    if(promise2 === x){//不能自己等待自己完成<br>        return reject(new TypeError(‘循环引用’));<br>    };<br>    // x是除了null以外的对象或者函数<br>    if(x !=null &amp;&amp; (typeof x === ‘object’ || typeof x === ‘function’)){<br>        let called;//防止成功后调用失败<br>        try{//防止取then是出现异常  object.defineProperty<br>            let then = x.then;//取x的then方法 {then:{}}<br>            if(typeof then === ‘function’){//如果then是函数就认为他是promise<br>                //call第一个参数是this，后面的是成功的回调和失败的回调<br>                then.call(x,y =&gt; {//如果Y是promise就继续递归promise<br>                    if(called) return;<br>                    called = true;<br>                    resolvePromise(promise2,y,resolve,reject)<br>                },r =&gt; { //只要失败了就失败了<br>                    if(called) return;<br>                    called = true;<br>                    reject(r);<br>                });<br>            }else{//then是一个普通对象，就直接成功即可<br>                resolve(x);<br>            }<br>        }catch (e){<br>            if(called) return;<br>            called = true;<br>            reject(e)<br>        }<br>    }else{//x = 123 x就是一个普通值 作为下个then成功的参数<br>        resolve(x)<br>    }</p>
<p>}
```</p>
<ul>
<li><p>返回结果可能为一个普通值，则直接   resolve(x);<br>Promise一次只能调用成功或者失败<br>也就是当调用成功就不能再调用失败了，如果两个都调用的时候，哪个先调用就执行哪一个。代码部分还是上面那部分</p>
</li>
<li><p>个人认为，这个地方比较绕，需要慢慢的一步一步的理清楚。</p>
</li>
<li><p>根据promise A+规范原理，promise在自己的框架中，封装了一系列的内置的方法。</p>
</li>
</ul>
<p>捕获错误的方法 catch()<br>解析全部方法 all()<br>竞赛 race()<br>生成一个成功的promise  resolve()<br>生成一个失败的promise  reject()</p>

    </div>
</div></div>
    </div>
    <div id="footer">不悔梦归处 只恨太匆匆</div>
    <img id="top" src="/images/up-circle.svg" alt="top">
    
  </body>
</html>
<!-- 加载主题脚本文件 -->
<script src="/scripts/utone.js"></script>

<script type="text/javascript">
  window.onload = function() {
    scroll.init()
    siteSearch.init()
  }
</script>
